### ISSUE TITLE
Crash within Tealium core android library for certain users

### ISSUE THREAD

**Issue by bc-bane** [Tealium/tealium-react-native#134]
Certain on our users are experiencing a crash on app startup that based on our logs seems to be an issue in the com.tealium.core.ModuleManager, we have seen it occur on android versions 9-13, but not for every android user or session. It seems to be an intermittent issue. Here is the crash log for the issue. 

`Fatal Exception: java.util.ConcurrentModificationException
       at java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:760)
       at java.util.LinkedHashMap$LinkedValueIterator.next(LinkedHashMap.java:787)
       at kotlin.collections.CollectionsKt___CollectionsJvmKt.filterIsInstanceTo(CollectionsKt___CollectionsJvmKt.java:35)
       at kotlin.collections.CollectionsKt___CollectionsJvmKt.filterIsInstance(CollectionsKt___CollectionsJvmKt.java:25)
       at com.tealium.core.ModuleManager.getModulesForType(ModuleManager.java:2)
       at com.tealium.core.ModuleManager.getModule(ModuleManager.java:1)
       at com.tealium.remotecommanddispatcher.RemoteCommandDispatcherKt.getRemoteCommands(RemoteCommandDispatcherKt.java:2)
       at com.tealium.react.TealiumReact.addRemoteCommand(TealiumReact.kt:252)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.facebook.react.bridge.JavaMethodWrapper.invoke(JavaMethodWrapper.java:372)
       at com.facebook.react.bridge.JavaModuleWrapper.invoke(JavaModuleWrapper.java:188)
       at com.facebook.react.bridge.queue.NativeRunnable.run(NativeRunnable.java)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:27)
       at android.os.Looper.loopOnce(Looper.java:226)
       at android.os.Looper.loop(Looper.java:313)
       at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run(MessageQueueThreadImpl.java:228)
       at java.lang.Thread.run(Thread.java:1012)`

**Comment by bc-bane** [comment]
Using tealium-react-native version 2.2.1, build.gradle file shows `implementation 'com.tealium:kotlin-core:1.3.0'`

**Comment by jameskeith** [comment]
Thanks @bc-bane - I can see where the issue would be coming from so I'll raise a ticket internally to have this addressed.

It looks like it's a timing issue with adding a new RemoteCommand whilst the Tealium instance is still initialising - we'll get a fix for this - but as a workaround, you might be able to avoid this by adding any RemoteCommand's to the `TealiumConfig` object instead of adding them separately:
```javascript
let config: TealiumConfig = {
    account: 'tealiummobile',
    profile: 'demo',
    environment: TealiumEnvironment.dev,
    /* ... keys removed ... */
    remoteCommands: [{
        id: "my-remote-command",
        path: "mappings.json", // optional path to mappings file,
        url: "https://mysite.com/mappings.json", // optional url to remote mappings file
        callback: (payload) => {
            // handle remote command
        }
    }]
};
```
Adding them like this should mean they are only added to the Tealium instance once it's ready

**Comment by bc-bane** [comment]
Thanks, I'll look into adjusting the way our remote commands are set up.

**Comment by acidealista** [comment]
@jameskeith has this issue been solved?
If yes, from which version was the fix introduced?
Thanks!

**Comment by mcapocaccia** [comment]
Hi @jameskeith we also have the issue and unfortunately it has multiplied by 5 times our crash rate on Android.

Can you please clarify if removing Dispatchers.RemoteCommands from the "dispatchers" key in the config is expected to solve the problem in v2.1.3? 

Or does a later version already include the fix?

Thank you in advance!

**Comment by jameskeith** [comment]
@acidealista 
I had assumed that since you had responded on this specific issue (related to a crash whilst adding Remote Commands) that this was also your issue.  

If that's not the case, and perhaps your stack-trace is pretty similar, then I would guess that it would be related to attempting to access a specific piece of module functionality before the SDK is fully initialized. You shouldn't need to add an empty array for the `remoteCommands` property if you're not using them.

If you are calling any methods on the `Tealium` class outside of the scope of the `callback` in `Tealium.initialize` then try moving those method calls inside the scope of the callback to ensure that all modules are fully initialized.

In either case, the SDK is not expected to cause a crash and the same fix being prepared for this issue should resolve your crashes also.

**Comment by mcapocaccia** [comment]
@jameskeith 

oh that's it.
We did not do anything to prevent a Tealium.track to be called while Tealium.initialize was still working.

We'll fix by wrapping track calls with a check on callback outcome and retry later if not ready, but probably enqueueing calls should be done internally by the SDK.

**Comment by jameskeith** [comment]
@mcapocaccia - your feedback is much appreciated; does that mean you were able to resolve it? I believe the underlying kotlin SDK will queue events sent via `track(..)` until the SDK is completely ready. There are a couple of optional modules that do not do this, and perhaps we can look at a better queueing solution in this integration as a whole.

However, since it's a slightly different issue to the one noted on this case, i'd like to suggest getting the error/stacktrace from your own issue if possible?
I want to make sure that we are actually fixing the issue for yourselves as well as the original report on this case.
If needs be then we can facilitate getting the information via support@tealium.com if it can't be posted here.

**Comment by mcapocaccia** [comment]
@jameskeith please see stack trace below, it is very similar to the one noted above in this case.

I can confirm that the app with the workaround hasn't crashed yet with a 20% adoption.

To recap, our issue was that we have a Tealium module which:
- calls Tealium.initialize in the module code, outside any function (i.e. at first module loading)
- contains a track function which calls Tealium.track

 Due to dynamic module importing, the module loading was happening exactly at first use time, i.e. on the first call to "track", and the initialisation could be still happening when Tealium.track got called.


Now we condition tracking to successful initialization COMPLETION (marked in the inizialize callback) and simply call Tealium.track or setTimeout(track, 1000) to retry later.

<img width="880" alt="image" src="https://github.com/Tealium/tealium-react-native/assets/17249177/7cc3dd2d-8b65-4d70-b7c9-8c105fdf665a">

**Comment by mcapocaccia** [comment]
@jameskeith unfortunately the problem hasn't disappeared with the workaround described above.

Looking at the crash reports, it seems to just happen within the execution of the initialize call itself, even if not disturbed by further Tealium calls.

We'll try to understand more with additional logging, but we'll be waiting for an updated release on your side to address it.
Any update on expected availability?

**Comment by jameskeith** [comment]
This should be resolved from v2.3.1 and above.
